.github/workflows/reusable-build-docsite.yaml

```
@@ -0,0 +1,30 @@
+name: Build the Ansible docsite
+
+on:
+  workflow_call:
+
+env:
+  PIP_DISABLE_PIP_VERSION_CHECK: 1
+
+jobs:
+  build-docsite:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out the repository
+        uses: actions/checkout@v3
+      - name: Set up Python
+        uses: actions/setup-python@v3
+        with:
+          python-version: "3.10"
+      - name: Install requirements
+        run: python -Im pip install nox
+      - name: Generate assets for the docs landing page
+        env:
+          PYTHONWARNINGS: error
```

@oraNod as for the CI erroring out: you can make another constraints file to limit the version of `setuptools` below v54.1.0 (might need to go further in the past). You can then integrate that file using an environment variable like `PIP_CONSTRAINT=build-deps.txt`. Using `-c` will not work for the build envs.

---

Hey @webknjaz Thanks for that. Good to know you've figured out it was the version of `setuptools`. I thought it was that but was really wrangling with how to overcome it.

Instead I would like to go with a simpler approach to use the `nox` action.

---

@oraNod with that, just discard the strictness. The action just replaces installing nox itself which was never problematic — the failure was in the build env provisioning. If your goal is to ignore the warnings, dropping the env var would be enough. Though, you'll get a hard failure in about a month, once the setuptools turns into a traceback. Having warnings turned into errors allows you to catch and fix future problems before it's too late.

---

Hey @webknjaz after a couple of runs with `staticjinja` 5.0.0  there are still the warnings about the `pkg_resources` deprecation in setuptools.

Once the setuptools issue turns into a traceback would we get the hard failure regardless of whether we do `pip -Im nox -s build` or use the nox action?

---

> Once the setuptools issue turns into a traceback would we get the hard failure regardless of whether we do `pip -Im nox -s build` or use the nox action?

Nox action only `pip install`'s nox itself. This never generated those warnings. Having just `pip install` changes nothing short of the syntax for installing nox.

The warnings are coming from `pip install -r requirements.in` as one of the dependencies is not packaged well, which is what you have in the noxfile and it's executed via `nox -s build`. The first poorly packaged sdist we encountered was `easywatch` that we got as a dependency of `staticjinja`. With its new release, this dependency is no longer pulled in, so it can't generate those warnings as it's not being built anymore.

So now we're facing a new warning, this time in pip itself — the one installed into the internal nox-managed virtualenv for the session. This is because pip vendors a lot of things, including `setuptools`' `pkg_resources` which is no longer supposed to be used. We can't really wait for pip to get rid of it, but we can try to suppress this specific warning instead.

---

According to my local tests, this should be enough to make it work:
```suggestion
          PYTHONWARNINGS: >-  # pip vendors deprecated `setuptools`' `pkg_resources`
            error,once:pkg_resources is deprecated as an API.:DeprecationWarning:pip._internal.metadata.importlib._envs
```

---

Once again, thanks for the solid gold explanation @webknjaz 

What is the best way to trace the new warning back to pip itself? I compared the warning to some of the earlier ones and could see differences. But I wasn't able to confidently determine exactly where it was coming from and if the `easywatch` removal fully resolved the warnings.

---

The traceback shows what Python modules the control flow travels. So with
```python-traceback
  File "~/src/github/python/cpython/some-venv/lib/python3.11/site-packages/pip/_internal/metadata/importlib/_envs.py", line 111, in _find_eggs_in_dir
    from pip._vendor.pkg_resources import find_distributions
  File "~/src/github/python/cpython/some-venv/lib/python3.11/site-packages/pip/_vendor/pkg_resources/__init__.py", line 118, in <module>
    warnings.warn(
DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
```
It's visible in the file paths and module imports. I also happen to know how pip vendors stuff so it was rather obvious to me.

The `easywatch` log was different — pip was in the middle of the dependency resolution so the first part of the log was rather normal, it showed that it'd found some wheels. But then, when it found `easywatch`, there was no wheel for it, so it initialized the process of building one, made a temporary throw-away build venv (which is normal) and provisioned it. It then invoked the build in that env, and that's where it hit the setuptools warning. Having the env var set, Python turned the warning into an error in that subprocess and errored it out, which then bubbled up back to pip and terminated it with a failure too.
The associated traceback referenced files in that temporary build env, and the file paths were somewhere under `/tmp` + originated from `setuptools` installed there.

I can see how that may be unobvious to many people, but I personally am involved in the packaging ecosystem, so for me this is just another day. I may not perceive errors like the less experienced folks.

---

Awesome. Thanks again for taking the time to explain that. I thought perhaps you had a diagnostic tool or perhaps used an additional flag for debugging. I know you're very involved with the packaging so I appreciate getting to see how you interpret these things and to align my own understanding. It's super helpful.

---

_I am_ the diagnostic tool :smiling_imp: :laughing: